File: leetCode/binaryTree/Binary_Tree_Level_Order_Traversal_102/Main.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

import javax.swing.tree.TreeNode;

// good explanation - https://www.geeksforgeeks.org/?p=2686

/**
 * Time Complexity.
 * O(N),
 * since each node in the tree will be pushed into the queue exactly once,
 * where N is the total number of nodes in the tree.
 * 
 * Space Complexity.
 * O(N),
 * the size of the queue will be at most N because
 * each node will be pushed into the queue exactly once.
 * 
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        Queue<TreeNode> nodeQueue = new LinkedList<>();
        if(root != null) {
            nodeQueue.offer(root);
        }
        while(!nodeQueue.isEmpty()) {
            ArrayList<Integer> levelList = new ArrayList<>();
            int queueSize = nodeQueue.size();

            for(int i = 0; i < queueSize; i++) {
                TreeNode curr = nodeQueue.poll();
                levelList.add(curr.val);
                if(curr.left != null) {
                    nodeQueue.offer(curr.left);
                }
                if(curr.right != null) {
                    nodeQueue.offer(curr.right);
                }
            }
            // build res
            res.add(levelList);
        }
        return res;
    }
}

File: leetCode/binaryTree/Binary_Tree_Level_Order_Traversal_102/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        q = deque()
        if root:
            q.append(root)

        res = []
        while q:
            q_size = len(q)
            level = []
            for _ in range(q_size):
                curr = q.popleft()

                if curr.left:
                    q.append(curr.left)
                if curr.right:
                    q.append(curr.right)

                level.append(curr.val)
            res.append(level)
        return res
